### PR TITLE
Auto-dampening feedback fixes

### PR DESCRIPTION
**Pull request by autoSteve**
* Allow a periodically resetting generation sensor (this likely will not break anything)
* Allow Watt-hour, kilowatt-hour or megawatt-hour generation/site export entities

### CODE REVIEW

**Review by BJReplay — APPROVED** [review]

**Comment by autoSteve** [comment]
Commit [376718d](https://github.com/BJReplay/ha-solcast-solar/pull/327/commits/376718dce2d706f8af0967927cf5258664ab628d) should solve for this:

image

The thing is I have not been able to replicate generation deltas going negative in the simulator when using a total increasing sensor as reported by @spry-salt, so that is most odd.

What the code does do now, though, is where a `diff()` item is negative it is sent to zero, which should inherently solve for and exclude negative. Further, it examines all of the usual tiny generation "bump" deltas and zeros anything over the 90th percentile for consideration by the auto-dampening model. Gaps in generation updates should already be handled.

So... I think that:

* Jumps in generation are covered
* Gaps in generation are covered
* Weird negative deltas _should_ be covered

These are now all excluded for implicated intervals under consideration by auto-dampening modelling. Bottom line: if it was a "bad day interval" then let that one go through to the keeper and consider all the other day intervals.

At this stage this is for generation only, and not the site export entity. That awaits a bloody good look in the same context of things "going away", and jumping about, as they do from time-to-time when poked.

**Comment by gcoan** [comment]
I'm unable to comment on the original issue as its locked now, but the readme changes look good @autoSteve 

One suggested minor reword above

**Comment by autoSteve** [comment]
In looking at the potential for site export entity anomalies causing havoc I have come to the conclusion that havoc will not ensue.

The worst that will happen for a "jump" is that a given generation interval where there is real export limiting occurring will likely not get picked up. That said, it would be equally likely at this anomalous time that PV generation entities would be similarly affected, given both numbers are probably coming from the same integration, so detection of limiting would be a moot point. Same outcome. No detected export limiting. The interval would not be considered for dampening.

Might be wrong, but I'm probably not given simulations.

> One suggested minor reword above

Can you recall what this suggestion was, @gcoan? I couldn't find your "above" reference.

**Comment by autoSteve** [comment]
I think this is pretty impressive for the afternoon, given today was variable, and the logarithmic adjustments were kicking in. Morning factors were all smashed to near 1.0 by log.

On the down-side I have an anomaly between 13:00-15:00 with four factors that have crept in recently. This is a totally shade-free time window, so a look at my own data is definitely needed. It did not cause undue forecast vs. generation difference because the factors are not "strong", but they are of interest nonetheless.

```
Morn, which is getting less shaded now. Very cool...

Granular dampening factor for 2025-09-12 07:00:00 is 0.917 (was 0.874, peak 1.258, interval pv50 0.898)
Granular dampening factor for 2025-09-12 07:30:00 is 0.912 (was 0.912, peak 2.193, interval pv50 2.292)
Granular dampening factor for 2025-09-12 08:00:00 is 0.919 (was 0.919, peak 2.937, interval pv50 3.009)

Arvo...

Odd...
Granular dampening factor for 2025-09-12 13:00:00 is 0.859 (was 0.786, peak 4.836, interval pv50 3.442)
Granular dampening factor for 2025-09-12 13:30:00 is 0.889 (was 0.813, peak 4.428, interval pv50 2.951)
Granular dampening factor for 2025-09-12 14:00:00 is 0.871 (was 0.720, peak 4.028, interval pv50 2.347)
Granular dampening factor for 2025-09-12 14:30:00 is 0.900 (was 0.864, peak 3.473, interval pv50 2.666)

Very cool...
Granular dampening factor for 2025-09-12 16:00:00 is 0.743 (was 0.643, peak 1.400, interval pv50 1.059)
Granular dampening factor for 2025-09-12 16:30:00 is 0.774 (was 0.489, peak 1.022, interval pv50 0.585)
Granular dampening factor for 2025-09-12 17:00:00 is 1.000 (was 0.317, peak 0.630, interval pv50 0.190)
```

image

(This is dampened forecast, BTW. I'll update with est. actuals once midnight rolls, which will be even more of interest.)

UPDATE: Dampened estimated actuals for the same day.

image

**Comment by gcoan** [comment]
> > One suggested minor reword above
>
> Can you recall what this suggestion was, @gcoan? I couldn't find your "above" reference.

It appears as a pending change from the PR review.

Here's a screenshot of it so you can see the original text (for context) and my suggestion:

image

**Comment by autoSteve** [comment]
> Here's a screenshot of it so you can see the original text (for context) and my suggestion

Wow.

Feeling like a total Github fail right now.

I still could not find that, yet have made that change. Thanks, @gcoan.

**Comment by autoSteve** [comment]
Maybe it's because I've got an 'approving' review already in the bag and could merge this sucker right now if it wasn't a draft. Maybe Github is being 'kind' to developers and masking other reviews? 🤷 I got nothing, and can find nothing, and can't add you as a reviewer because you're not set up as a collab.

I am refusing to question anything on the Internet, or further in life right now. I have a brilliant Shinobu 15 y/o Japanese whisky in hand in an epically weighty hand-crafted Japanese cut crystal Kagami whisky glass, and shall ease into the rest of Friday night without a care, knowing that over ten thousand souls have not opened a Github issue for v4.4.0. (Touches wood..)

**Comment by autoSteve** [comment]
Yeah, marked as ready for review, @BJReplay because no one else so far has needed the Mercurochrome after a 'stack'.

**Comment by BJReplay** [comment]
> I got nothing, and can find nothing

I couldn't see it, either.

File: tests/__init__.py [hunk 1]
```diff
@@ -299,46 +312,108 @@ async def async_setup_aioresponses() -> None:
...
+            _uom = "kWh"
+        case ExtraSensors.YES_NO_UNIT:
+            _uom = ""
+        case ExtraSensors.DODGY:
+            _uom = "MJ"
+        case _:
+            _uom = "kWh"
+    _DAYS = 1
+
+    adjustment = {"kWh": 1.0, "MWh": 1000.0, "Wh": 0.001, "MJ": 1.0, "": 1.0}
```
Comment (BJReplay) [inline tests/__init__.py:333]: Dodgy adjustment for dodgy factor 🤣

Comment (autoSteve) [reply tests/__init__.py:333]: Tee hee.

**Comment by BJReplay** [comment]
Ready to go, I reckon, and do the templates separately, @autoSteve & @Nilogax

**Comment by autoSteve** [comment]
I reckon this wants more thought, @BJReplay...

My middle-of-day factors that are being introduced are on the back of a single generation interval.

I think this is likely because these single values are from a day where my solar export was _not_ curtailed and it was a reasonably good day. These are incorrect.

Way back I was requiring to have the same number of generation values as "good day" actuals, and it was too aggresive in certain circumstances. Maybe there should be a "majority" of generation points (i.e >3 for 13:00, >2 for 13:30)?

I will code that and see how it runs.

```
Interval 13:00 has peak estimated actual 4.836 and 6 matching intervals: 09-01, 09-05, 09-06, 09-08, 09-11, 09-12
Interval 13:00 max generation: 3.800, [3.8]
Auto-dampen factor for 13:00 is 0.786
Interval 13:30 has peak estimated actual 4.428 and 4 matching intervals: 09-05, 09-06, 09-09, 09-11
Interval 13:30 max generation: 3.600, [3.6]
Auto-dampen factor for 13:30 is 0.813
Interval 14:00 has peak estimated actual 4.017 and 4 matching intervals: 09-05, 09-06, 09-09, 09-11
Interval 14:00 max generation: 2.900, [2.9]
Auto-dampen factor for 14:00 is 0.722
Interval 14:30 has peak estimated actual 3.463 and 4 matching intervals: 09-05, 09-06, 09-09, 09-11
Interval 14:30 max generation: 3.000, [3.0]
Auto-dampen factor for 14:30 is 0.866
```

**Comment by autoSteve** [comment]
Yeah. Much better...

image

**Comment by autoSteve** [comment]
`pytest` code needs a tweak now.

**Comment by BJReplay** [comment]
> Yeah. Much better...

At least you exceed your 10% / 90% POE - me - not so much:

image

I'm likely to be below my "don't dump to Zero Hero" threshold of 85% by 6pm.

**Comment by autoSteve** [comment]
> `pytest` code needs a tweak now.

`pytest` code is tweaked.

It is also performance optimised in some `__init__` places that are "expensive" to run.

Test runtime is the same as it was, despite building more data. It's slow-ish for auto-dampen, but the code is tested properly with integration tests, and not just unit tests. Entity history actually gets built using synthetic time, and that is expensive to do for entities that "jump" by 0.1 kWh every so often.

I think this code is now solid and 'sendable'.

The example in `TEMPLATES.md` will not need a release because `README.md` is not manhandled, so that can be merged any time.

**Comment by autoSteve** [comment]
I'm liking the "blue sky" dampening experience from this arvo. Morning self-shading from the gabled roof continues to reduce every day. New data points unsurprisingly take a day to be reflected in the modelling, which can't be helped and is inherent by design.

With a boosted model it may be possible to anticipate new data points for lengthening / shortening days, with data points "stretched" or "compressed" by some inform but let's not go there.

This feature is just complicated enough to deserve a chunky readme section as is, which has (gah!) already been reported as too hard to understand.

It's a hobbyist forecast, so this is not getting any more complicated than the simple model we have that appears to work really quite well.

image

**Comment by autoSteve** [comment]
I feel inclined to re-model the readme section for automated dampening.

There is nothing that is technically incorrect, but the order of information bothers me.

The TL>DR comes far too late, which is the "theory of operation". That deserves pride of place, and then those that do not understand the component parts can read on. Or raise an issue because they didn't read on.

"Be kind to your reader", and all that...

**Comment by BJReplay** [comment]
Readme looks good (and better, now), but that assumes that people read the readme.

To the extent that prompting people is a thing, perhaps the release notes should have:

# **Breaking Change**
## **_Read the Readme_**
No, there's not actually a breaking change, but lots has changed since earlier releases, and if you don't read the readme, and then post issues or discussions, you'll be pointed back to the readme, so please read the readme before installing this version or setting up automated dampening.

Dampening much improved today:

image

**Review by BJReplay — APPROVED** [review]
I am away for work again until Friday from early tomorrow.

LGTM.

Feel free to send it when you are confident, and ping me to promote release discussion to announcement.

**Comment by gcoan** [comment]
Appreciate you have just approved this @BJReplay but I got pinged with the notification of latest readme changes so I read them..

@autoSteve This bit:

"If it departs greatly from best then a significant change is made to the factor to render it ineffective (i.e. closer to 1.0 x forecasted generation). This determination is made based on the value of every forecasted interval.

The rendering of ineffective for factors is because heavily clouded intervals are likely to have diffuse light as the most significant generation component, and not direct sunlight, which is the component most impacted by shade."

I had to read three times to understand it and wonder if it could be reworded?

'significant change' to me reads as you're making a big adjustment to the dampening factor, but then rest of the sentence says its moved to be closer to 1.0 which sounds the opposite of big adjustment.
The use of the phrase 'factors' more than once also doesn't aid readability, suggest using the full 'dampening factors'?

So rather than just throwing rocks, here's my suggested reword.  More words but shorter sentences with less concepts in each sentence:

"If the actual solar generation varies significantly to the 'best' estimated solar generation then the dampening factor is adjusted so it has little impact (i.e. closer to 1.0 x forecasted generation). This determination is made based on the value of every forecasted interval.

This dampening factor adjustment is because significant solar generation variance is indicative of heavily clouded intervals. Such clouded intervals are likely to have diffuse light as the most significant component of solar generation, and not direct sunlight, which is the solar generation component most impacted by shade."

**Comment by BJReplay** [comment]
Thanks @gcoan.

Another _suggested tweak_, borrowing from your work.

"If the _forecast_ solar generation varies significantly to the 'best' estimated solar generation _(e.g. it is very cloudy)_ then the dampening factor is adjusted so it has little impact (i.e. _adjusted_ closer to 1.0 x forecasted generation). This determination is made based on the value of every forecasted interval.

This dampening factor adjustment is because significant solar generation variance is indicative of heavily clouded intervals. Such clouded intervals are likely to have diffuse light as the most significant component of solar generation, and not direct sunlight, which is the solar generation component most impacted by shade."

**Comment by gcoan** [comment]
> Another _suggested tweak_, borrowing from your work.

looks good.   I thought the comparison was of actual to estimated generation, but you've seen the code more than me!  Shows the importance of being explicit and spelling everything out

so people can still not understand but can be pointed to the readme

there are a lot of concepts going on here

cheers

**Comment by autoSteve** [comment]
Where I landed is middle ground, plus a 'stretch goal' of a  bonus 'tip'.

**Comment by Nilogax** [comment]
Re-reading revised README (try saying that after a night out!) before updating my wording in the other PR and a couple of things jumped out.

In the preview view the screenshot of the config appears as a link rather than an image

" the dampening factor is adjusted so it has little impact (i.e. adjusted closer to 1.0 x forecasted generation)." might be confusing.  "(ie. adjusted closer to 1.0)" makes more sense to me - if forecasted generation is 2.3kWh then the dampening factor isn't adjusted closer to 1.0x2.3.

inconsistency of single and double quotes around "best"/'best'

Also FWIW you may recall I did a scatter plot comparing %(actual vs forecast) with %(forecast vs peak) to see if there was any discernable trend that might back up or disprove the logarithmic adjustment.  Well I updated the plot with some more data to check and.... nope.    I have however noticed a consistent optimism in dampened forecast vs actual (I've been logging %variance since i started ye olde dampening method) and am currently running with half strength logarithmic nudging.  Obviously that's bringing the forecast total down a bit so variance has decreased but it's too early to tell if it's a worthwhile change or just an artefact of reducing day length giving slightly optimistic data.  And yes, it's only a forecast.

**Comment by autoSteve** [comment]
Thanks for the proof read, @Nilogax.

**Comment by BJReplay** [comment]
> Thanks for the proof read, @Nilogax.

I've gotta say that I think that the explanation is getting more understandable through the collective efforts of the author and the reviews, so thanks, all.

**Comment by autoSteve** [comment]
FYI, @Nilogax, your PR example in `TEMPLATES.md` is now referenced in the `README.md`.

A few more other tweaks.

**Comment by autoSteve** [comment]
There's a few more tests now. 100% cover.

```
(ha-venv) vscode ➜ /workspaces/…/tests/components/solcast_solar $ pytest --cov=homeassistant.components.solcast_solar --cov-report term-missing -v
Test session starts (platform: linux, Python 3.13.7, pytest 8.4.2, pytest-sugar 1.0.0)
cachedir: .pytest_cache
rootdir: /workspaces/homeAssistant-core
configfile: pyproject.toml
plugins: sugar-1.0.0, pytest_freezer-0.4.9, respx-0.22.0, github-actions-annotate-failures-0.3.0, timeout-2.4.0, socket-0.7.0, xdist-3.8.0, picked-0.5.1, syrupy-4.9.1, aiohttp-1.1.0, asyncio-1.1.0, requests-mock-1.12.1, cov-6.2.1, unordered-0.7.0, anyio-4.10.0
asyncio: mode=Mode.AUTO, asyncio_default_fixture_loop_scope=function, asyncio_default_test_loop_scope=function
collected 98 items                                                                                                                                                                                                                               

 tests/components/solcast_solar/test_config_flow.py::test_single_instance ✓                                                                                                                                                          1% ▏         
 tests/components/solcast_solar/test_config_flow.py::test_create_entry ✓                                                                                                                                                             2% ▎         
 tests/components/solcast_solar/test_config_flow.py::test_init_api_key[user_input0-api_looks_like_site] ✓                                                                                                                            3% ▍         
 tests/components/solcast_solar/test_config_flow.py::test_init_api_key[user_input1-api_duplicate] ✓                                                                                                                                  4% ▌         
 tests/components/solcast_solar/test_config_flow.py::test_init_api_key[user_input2-None] ✓                                                                                                                                           5% ▌         
 tests/components/solcast_solar/test_config_flow.py::test_init_api_key[user_input3-None] ✓                                                                                                                                           6% ▋         
 tests/components/solcast_solar/test_config_flow.py::test_init_api_key[user_input4-None] ✓                                                                                                                                           7% ▊         
 tests/components/solcast_solar/test_config_flow.py::test_init_api_key[user_input5-limit_one_or_greater] ✓                                                                                                                           8% ▊         
 tests/components/solcast_solar/test_config_flow.py::test_config_api_key_invalid ✓                                                                                                                                                   9% ▉         
 tests/components/solcast_solar/test_config_flow.py::test_config_api_quota[options0-user_input0-limit_not_number] ✓                                                                                                                 10% █         
 tests/components/solcast_solar/test_config_flow.py::test_config_api_quota[options1-user_input1-limit_one_or_greater] ✓                                                                                                             11% █▎        
 tests/components/solcast_solar/test_config_flow.py::test_config_api_quota[options2-user_input2-limit_too_many] ✓                                                                                                                   12% █▎        
 tests/components/solcast_solar/test_config_flow.py::test_config_api_quota[options3-user_input3-None] ✓                                                                                                                             13% █▍        
 tests/components/solcast_solar/test_config_flow.py::test_config_api_quota[options4-user_input4-None] ✓                                                                                                                             14% █▌        
 tests/components/solcast_solar/test_config_flow.py::test_config_api_quota[options5-user_input5-limit_too_many] ✓                                                                                                                   15% █▌        
 tests/components/solcast_solar/test_config_flow.py::test_config_api_quota[options6-user_input6-None] ✓                                                                                                                             16% █▋        
 tests/components/solcast_solar/test_config_flow.py::test_reauth_api_key[component.solcast_solar.config.error.Bad API key, 403/Forbidden returned for ******555] ✓                                                                  17% █▊        
 tests/components/solcast_solar/test_config_flow.py::test_reconfigure_api_key1 ✓                                                                                                                                                    18% █▉        
 tests/components/solcast_solar/test_config_flow.py::test_reconfigure_api_key2[None-options0-Bad API key, 403/Forbidden-ignore_missing_translations0] ✓                                                                             19% █▉        
 tests/components/solcast_solar/test_config_flow.py::test_reconfigure_api_key2[None-options1-No sites for the API key-ignore_missing_translations1] ✓                                                                               20% ██▏       
 tests/components/solcast_solar/test_config_flow.py::test_reconfigure_api_key2[return_429-options2-Error 429/Try again later for API key-ignore_missing_translations2] ✓                                                            21% ██▎       
 tests/components/solcast_solar/test_config_flow.py::test_reconfigure_api_key2[exception-options3-None-ignore_missing_translations3] ✓                                                                                              22% ██▎       
 tests/components/solcast_solar/test_config_flow.py::test_reconfigure_api_key2[None-options4-None-ignore_missing_translations4] ✓                                                                                                   23% ██▍       
 tests/components/solcast_solar/test_config_flow.py::test_reconfigure_api_quota ✓                                                                                                                                                   24% ██▌       
 tests/components/solcast_solar/test_config_flow.py::test_options_api_key[user_input0-api_looks_like_site] ✓                                                                                                                        26% ██▋       
 tests/components/solcast_solar/test_config_flow.py::test_options_api_key[user_input1-api_duplicate] ✓                                                                                                                              27% ██▋       
 tests/components/solcast_solar/test_config_flow.py::test_options_api_key[user_input2-None] ✓                                                                                                                                       28% ██▊       
 tests/components/solcast_solar/test_config_flow.py::test_options_api_key[user_input3-None] ✓                                                                                                                                       29% ██▉       
 tests/components/solcast_solar/test_config_flow.py::test_options_api_key[user_input4-None] ✓                                                                                                                                       30% ██▉       
 tests/components/solcast_solar/test_config_flow.py::test_options_api_key[user_input5-limit_one_or_greater] ✓                                                                                                                       31% ███▏      
 tests/components/solcast_solar/test_config_flow.py::test_options_api_key_invalid ✓                                                                                                                                                 32% ███▎      
 tests/components/solcast_solar/test_config_flow.py::test_options_api_quota[options0-user_input0-limit_not_number] ✓                                                                                                                33% ███▍      
 tests/components/solcast_solar/test_config_flow.py::test_options_api_quota[options1-user_input1-limit_one_or_greater] ✓                                                                                                            34% ███▍      
 tests/components/solcast_solar/test_config_flow.py::test_options_api_quota[options2-user_input2-limit_too_many] ✓                                                                                                                  35% ███▌      
 tests/components/solcast_solar/test_config_flow.py::test_options_api_quota[options3-user_input3-None] ✓                                                                                                                            36% ███▋      
 tests/components/solcast_solar/test_config_flow.py::test_options_api_quota[options4-user_input4-None] ✓                                                                                                                            37% ███▋      
 tests/components/solcast_solar/test_config_flow.py::test_options_api_quota[options5-user_input5-limit_too_many] ✓                                                                                                                  38% ███▊      
 tests/components/solcast_solar/test_config_flow.py::test_options_api_quota[options6-user_input6-None] ✓                                                                                                                            39% ███▉      
 tests/components/solcast_solar/test_config_flow.py::test_options_custom_hour_sensor[options0-0-custom_invalid] ✓                                                                                                                   40% ████      
 tests/components/solcast_solar/test_config_flow.py::test_options_custom_hour_sensor[options1-145-custom_invalid] ✓                                                                                                                 41% ████▏     
 tests/components/solcast_solar/test_config_flow.py::test_options_custom_hour_sensor[options2-8-None] ✓                                                                                                                             42% ████▎     
 tests/components/solcast_solar/test_config_flow.py::test_options_hard_limit[options0-invalid-hard_not_number] ✓                                                                                                                    43% ████▍     
 tests/components/solcast_solar/test_config_flow.py::test_options_hard_limit[options1--1-hard_not_number] ✓                                                                                                                         44% ████▍     
 tests/components/solcast_solar/test_config_flow.py::test_options_hard_limit[options2-6,6.0-hard_too_many] ✓                                                                                                                        45% ████▌     
 tests/components/solcast_solar/test_config_flow.py::test_options_hard_limit[options3-6-None] ✓                                                                                                                                     46% ████▋     
 tests/components/solcast_solar/test_config_flow.py::test_options_hard_limit[options4-6,6.0-None] ✓                                                                                                                                 47% ████▊     
 tests/components/solcast_solar/test_config_flow.py::test_options_hard_limit[options5-6-None] ✓                                                                                                                                     48% ████▊     
 tests/components/solcast_solar/test_config_flow.py::test_options_hard_limit[options6-0-None] ✓                                                                                                                                     49% ████▉     
 tests/components/solcast_solar/test_config_flow.py::test_options_auto_dampen[options0-actuals_without_get] ✓                                                                                                                       50% █████     
 tests/components/solcast_solar/test_config_flow.py::test_options_auto_dampen[options1-dampen_without_actuals] ✓                                                                                                                    51% █████▏    
 tests/components/solcast_solar/test_config_flow.py::test_options_auto_dampen[options2-dampen_without_generation] ✓                                                                                                                 52% █████▎    
 tests/components/solcast_solar/test_config_flow.py::test_options_auto_dampen[options3-export_multiple_entities] ✓                                                                                                                  53% █████▍    
 tests/components/solcast_solar/test_config_flow.py::test_options_auto_dampen[options4-export_no_entity] ✓                                                                                                                          54% █████▌    
 tests/components/solcast_solar/test_config_flow.py::test_step_to_dampen ✓                                                                                                                                                          55% █████▌    
 tests/components/solcast_solar/test_config_flow.py::test_dampen[value0] ✓                                                                                                                                                          56% █████▋    
 tests/components/solcast_solar/test_config_flow.py::test_entry_options_upgrade ✓                                                                                                                                                   57% █████▊    
 tests/components/solcast_solar/test_config_flow.py::test_presumed_dead_and_full_flow ✓                                                                                                                                             58% █████▊    
 tests/components/solcast_solar/test_dampen.py::test_auto_dampen ✓                                                                                                                                                                  59% █████▉    
 tests/components/solcast_solar/test_dampen.py::test_auto_dampen_issues[ExtraSensors.YES_UNIT_NOT_IN_HISTORY] ✓                                                                                                                     60% ██████    
 tests/components/solcast_solar/test_dampen.py::test_auto_dampen_issues[ExtraSensors.YES_NO_UNIT] ✓                                                                                                                                 61% ██████▎   
 tests/components/solcast_solar/test_dampen.py::test_auto_dampen_issues[ExtraSensors.DODGY] ✓                                                                                                                                       62% ██████▎   
 tests/components/solcast_solar/test_dampen.py::test_percentile ✓                                                                                                                                                                   63% ██████▍   
 tests/components/solcast_solar/test_diagnostics.py::test_diagnostics ✓                                                                                                                                                             64% ██████▌   
 tests/components/solcast_solar/test_energy.py::test_energy_data ✓                                                                                                                                                                  65% ██████▌   
 tests/components/solcast_solar/test_forecast_retry.py::test_forecast_retry ✓                                                                                                                                                       66% ██████▋   
 tests/components/solcast_solar/test_integration.py::test_api_failure ✓                                                                                                                                                             67% ██████▊   
 tests/components/solcast_solar/test_integration.py::test_schema_upgrade ✓                                                                                                                                                          68% ██████▉   
 tests/components/solcast_solar/test_integration.py::test_integration[options0] ✓                                                                                                                                                   69% ██████▉   
 tests/components/solcast_solar/test_integration.py::test_integration[options1] ✓                                                                                                                                                   70% ███████▏  
 tests/components/solcast_solar/test_integration.py::test_integration[options2] ✓                                                                                                                                                   71% ███████▎  
 tests/components/solcast_solar/test_integration.py::test_integration[options3] ✓                                                                                                                                                   72% ███████▎  
 tests/components/solcast_solar/test_integration.py::test_remaining_actions ✓                                                                                                                                                       73% ███████▍  
 tests/components/solcast_solar/test_integration.py::test_scenarios ✓                                                                                                                                                               74% ███████▌  
 tests/components/solcast_solar/test_integration.py::test_estimated_actuals ✓                                                                                                                                                       76% ███████▋  
 tests/components/solcast_solar/test_repairs.py::test_missing_data_fixable ✓                                                                                                                                                        77% ███████▋  
 tests/components/solcast_solar/test_repairs.py::test_unusual_azimuth[scenario0] ✓                                                                                                                                                  78% ███████▊  
 tests/components/solcast_solar/test_repairs.py::test_unusual_azimuth[scenario1] ✓                                                                                                                                                  79% ███████▉  
 tests/components/solcast_solar/test_repairs.py::test_unusual_azimuth[scenario2] ✓                                                                                                                                                  80% ███████▉  
 tests/components/solcast_solar/test_repairs.py::test_unusual_azimuth[scenario3] ✓                                                                                                                                                  81% ████████▏ 
 tests/components/solcast_solar/test_repairs.py::test_unusual_azimuth[scenario4] ✓                                                                                                                                                  82% ████████▎ 
 tests/components/solcast_solar/test_repairs.py::test_unusual_azimuth[scenario5] ✓                                                                                                                                                  83% ████████▍ 
 tests/components/solcast_solar/test_repairs.py::test_unusual_azimuth[scenario6] ✓                                                                                                                                                  84% ████████▍ 
 tests/components/solcast_solar/test_repairs.py::test_unusual_azimuth[scenario7] ✓                                                                                                                                                  85% ████████▌ 
 tests/components/solcast_solar/test_repairs.py::test_unusual_azimuth[scenario8] ✓                                                                                                                                                  86% ████████▋ 
 tests/components/solcast_solar/test_repairs.py::test_unusual_azimuth[scenario9] ✓                                                                                                                                                  87% ████████▋ 
 tests/components/solcast_solar/test_repairs.py::test_unusual_azimuth[scenario10] ✓                                                                                                                                                 88% ████████▊ 
 tests/components/solcast_solar/test_repairs.py::test_unusual_azimuth[scenario11] ✓                                                                                                                                                 89% ████████▉ 
 tests/components/solcast_solar/test_select.py::test_select_change_value[0-estimate-forecast_today-42.552] ✓                                                                                                                        90% █████████ 
 tests/components/solcast_solar/test_select.py::test_select_change_value[1-estimate10-forecast_today-35.46] ✓                                                                                                                       91% █████████▏
 tests/components/solcast_solar/test_select.py::test_select_change_value[2-estimate90-forecast_today-47.28] ✓                                                                                                                       92% █████████▎
 tests/components/solcast_solar/test_sensor.py::test_sensor_states[2-settings0] ✓                                                                                                                                                   93% █████████▍
 tests/components/solcast_solar/test_sensor.py::test_sensor_states[1-settings1] ✓                                                                                                                                                   94% █████████▍
 tests/components/solcast_solar/test_sensor.py::test_sensor_x_hours_long ✓                                                                                                                                                          95% █████████▌
 tests/components/solcast_solar/test_sensor.py::test_sensor_unavailable ✓                                                                                                                                                           96% █████████▋
 tests/components/solcast_solar/test_sensor.py::test_sensor_unavailable_exception ✓                                                                                                                                                 97% █████████▊
 tests/components/solcast_solar/test_specific_times.py::test_midnight ✓                                                                                                                                                             98% █████████▊
 tests/components/solcast_solar/test_specific_times.py::test_timezone_transition ✓                                                                                                                                                  99% █████████▉
 tests/components/solcast_solar/test_system_health.py::test_system_health ✓                                                                                                                                                        100% ██████████
================================================================================================================= tests coverage =================================================================================================================
________________________________________________________________________________________________ coverage: platform linux, python 3.13.7-final-0 _________________________________________________________________________________________________

Name                                                                                     Stmts   Miss  Cover   Missing
----------------------------------------------------------------------------------------------------------------------
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/__init__.py          467      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/config_flow.py       297      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/const.py              53      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/coordinator.py       426      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/diagnostics.py        17      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/energy.py             16      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/repairs.py            41      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/select.py             42      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/sensor.py            162      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/solcastapi.py       1834      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/system_health.py      11      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/util.py              149      0   100%
----------------------------------------------------------------------------------------------------------------------
TOTAL                                                                                     3515      0   100%

Results (59.86s):
      98 passed
```

**Comment by autoSteve** [comment]
`pytest` is expanded and 💯 (sez me... right...) ☑. v4.4.1 release is drafted ☑, and in the tubes awaiting merge of this lot. `manifest` updated ☑. `readme` updated ☑. Fingers crossed ☑.

I might send this early as a pre-weekend present, @BJReplay, depending on further chat about the registry and @webwude's sensor that is a sensor, yet is not a sensor in the entity registry. Fascinating.

**Comment by autoSteve** [comment]
A late breaking fix.

I accidentally published my API key tonight, so then I changed my API key, thinking, "This'll be a good test". And it was. This should not have resulted in an API key usage tracking reset because nothing else changed, yet it did result in a reset.

This is now fixed, @BJReplay. What's in an indent, eh? Yeah, significance.

This still slips through `pytest` code, but it is such a corner case that my giveadamn is kind of busted. The tests do check for reset on sites changed in conjunction with an API key change, which is likely a wholesale change of setup, and that's the one where API usage should get reset. The benign change use case is, well, benign, and checks out in the real world now.

If I were truly AR I would build a specific test, and I may one day. It's on the list.

That or it'll be committed soon. Who can say? It may just be that I am completely AR and can't let it go...

**Comment by BJReplay** [comment]
> I might send this early as a pre-weekend present

Lemme know if you do so I can promote the announcement; back home, but off to the sack soonish, @autoSteve.

**Comment by autoSteve** [comment]
I may, or may not be AR... 😅

Test added.

I need to put this down and think of something new to do. Gonna send it.

"Something new" is most likely going to be detection of Amber and similar users avoiding any export when wholesale goes negative, which would probably throw off auto-dampen in times of regular Summer sun-shinyness. That's kind-of like an export limit scenario, but in that case PV production will equal site usage with zero site export.

The simple answer to that is to add a site usage entity into the mix. The harder answer is, "Are these entity values actually going to match, or is there going to be a slight real-world measurement difference between Enphase and Tesla? There is for me... yet the options dialogue grows.

I do so wish that HA would add the ability to simply split out an "advanced" config thingamy, and not just a step from one dialogue in the flow to another.

**Comment by Nilogax** [comment]
> I do so wish that HA would add the ability to simply split out an "advanced" config thingamy

FWIW Predbat now has a Web based front end with dashboard, access to logs, editable config etc.   I reckon that would keep you busy for a while 😁

**Comment by autoSteve** [comment]
If I could laugh harder in github I would, @Nilogax. 😘

**Comment by autoSteve** [comment]
It is done.

Announcement is up, @BJReplay.
